Add support for AMD SGPIO enclosure management

nfont opened the pull request:
Add support for SGPIO register handling on AMD systems with enclosure
management support.

Patch 1/2 splits the vendor id check out of the generic AHCI
check to allow for AHCI checks for different vendors.

Patch 2/2 introduces support to manage LED settings via SGPIO
for AMD systems.
---

Nathan Fontenot (2):
      Split the vendor id check out of ahci check.
      Add support for AMD SGPIO enclosures


 src/Makefile    |    3 
 src/amd_sgpio.c |  762 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/amd_sgpio.h |   24 ++
 src/block.c     |    5 
 src/cntrl.c     |   27 ++
 src/cntrl.h     |    3 
 6 files changed, 820 insertions(+), 4 deletions(-)
 create mode 100644 src/amd_sgpio.c
 create mode 100644 src/amd_sgpio.h

--
Nathan Fontenot

mtkaczyk commented on src/amd_sgpio.c:
```diff
@@ -0,0 +1,762 @@
+} __attribute__ ((__packed__));
+
+struct sgpio_amd_register {
+	uint8_t		initiator:1;
```
The binary fields are not designed for mapping registers. Here I found nice explanation: 
https://stackoverflow.com/questions/25345691/bit-fields-portability
So please use bitmasks.
I know that ledmon has this kind of mapping for SCSI in smp.c. It will be changed soon.

mtkaczyk commented on src/amd_sgpio.c:
```diff
@@ -0,0 +1,762 @@
+		if (count == reg_len || errno != EBUSY)
+			break;
+
+		sleep(1);
```
If it is possible please use nanosleep with smaller value.
If not leave it as is.

apaszkie commented on src/amd_sgpio.c:
```diff
@@ -51,49 +51,49 @@
-	uint8_t		return_to_normal:1;
-	uint8_t		unused:1;
-	uint8_t		reserved2[3];
+	uint32_t	initiator:1;
```
Let me clarify this a little. The C Standard states: "The order of allocation of bit-fields within a unit (high-order to low-order or low-order to high-order) is implementation-defined." ([http://c0x.coding-guidelines.com/6.7.2.1.html#1412](url))

So there is no guarantee that this code will work as expected on a different compiler. Please use shift & mask if you must access specific bits from hardware.

mtkaczyk commented on src/amd_sgpio.c:
```diff
@@ -0,0 +1,762 @@
+	 *	cache_entry[n] => drive (4*n) to drive (4*n + 3)
+	 */
+	base_port = (drive->ata_port / 4) * 4;
+	index = base_port / 4;
```
You missed this one.

mtkaczyk commented on src/amd_sgpio.c:
```diff
@@ -0,0 +1,907 @@
+		return 0;
+	}
+
+	/* Validate that enlosure management is supported */
```
"enclosure"

mtkaczyk commented on src/amd_sgpio.c:
```diff
@@ -444,8 +445,10 @@ static int _find_file_path(const char *start_path, const char *filename,
+			char tmp[PATH_MAX];
+
+			strncpy(tmp, dir_path, path_len);
+			snprintf(path, path_len, "%s", dirname(tmp));
```
Second missed comment.

nfont commented:
> Hi,
> It looks good, but you missed two comments, and I found one typo.
> Please resolve them.
> After that I will merge it.
> 
> Thanks,
> Mariusz

For the first comment you pointed out, I have updated the wording and removed the cache_entry[x].. pieces. Was there anything else you were looking to add/change/remove from the comment?
